File: py/pe/pe4.py
#!/usr/bin/env python
"""
Solve project euler 4

Find the largest palindrome made from the product of two 3-digit numbers.
"""

from utils import is_palindrome

def pe4(n=3):
    """
    >>> pe4()
    906609
    """
    first, last = 9*10**(n - 1) + 1, 10**n
    mx = 0
    for x in range(first, last):
        for y in range(x, last):
            xy = x * y
            if xy > mx and is_palindrome(xy):
                mx = xy
    return mx

if __name__ == "__main__":
    import doctest
    doctest.testmod()

    try:
        while True:
            s = input('> ')
            n = int(s)
            print(pe4(n))
    except (SyntaxError, EOFError, KeyboardInterrupt, NameError):
        pass

File: cpp/pe/src/pe4.cc
#include "../include/pe4.h"
#include "../include/palindrome.h"

#include <cmath>

namespace pe {
  uint32_t pe4(uint32_t n, uint32_t& first, uint32_t& last)
  {
    first = static_cast<uint32_t>(9 * pow(10, n - 1) + 1);
    last = static_cast<uint32_t>(pow(10, n));

    uint32_t m = 0;
    for (uint32_t x = first; x < last; ++x) {
      for (uint32_t y = x; y < last; ++y) {
        uint32_t xy = x * y;
        if (xy > m && is_palindrome(xy)) {
          m = xy;
        }
      }
    }

    return m;
  }
}